### DISCUSSION TITLE
Poor accuracy of `GridInterpolationKernel` on 2D grid with different distances along each axis

### DISCUSSION THREAD

**Discussion by Edenhofer** [cornellius-gp/gpytorch#1996] [Q&A]
The following script demonstrates the poor accuracy of the `GridInterpolationKernel` on a 2D grid. The pixels of the first axis in the example are spaced two unit apart while the pixels of the second axis are spaced one unit apart. The covariance of the GPytorch GP does not compare favorably to the true covariance. For demonstration purposes I have not used `GridKernel` because I intend to extend this to an irregular grid.

The poor accuracy of the covariance got me by surprise. The underlying cause seems to be that the grid points of the interpolation kernel are laid out along the diagonal of the grid and no outer product of the two grid axes is performed to retrieve the full 2D grid. While this might save quite a bit of memory, it completely neglects the different scales of the two axis.

If my assertion of the underlying problem is correct, is it possible to have greater control over the layout of the inducing points for the grid interpolation as to avoid this diagonal layout?

```python
#!/usr/bin/env python3

from functools import partial

import gpytorch
from matplotlib import pyplot as plt
import matplotlib.pyplot as plt
import numpy as np
from scipy.spatial import distance_matrix
import torch

torch.set_default_dtype(torch.float64)


# %%
def matern_kernel(distance, scale, cutoff, dof):
    from scipy.special import gammaln, kv

    reg_dist = np.sqrt(2 * dof) * distance / cutoff
    cov = scale**2 * 2**(1 - dof) / np.exp(
        gammaln(dof)
    ) * (reg_dist)**dof * kv(dof, reg_dist)
    cov = np.where(reg_dist == 0., scale**2, cov)  # Fix limit
    return cov


def gpytorch_grid_interpolation_cov(scale, cutoff, dof, grid_size, grid_bounds):
    # HACK to fix the values of `cutoff` and `scale`; Is there a better way?
    cutoff_prior = gpytorch.priors.UniformPrior(
        a=cutoff, b=(1. + 1e-14) * cutoff
    )
    cutoff_interval = gpytorch.constraints.Interval(
        cutoff, (1. + 1e-7) * cutoff
    )
    scale_prior = gpytorch.priors.UniformPrior(
        a=scale**2, b=(1. + 1e-14) * scale**2
    )
    scale_interval = gpytorch.constraints.Interval(
        scale**2, (1. + 1e-7) * scale**2
    )
    gpy_kernel = gpytorch.kernels.MaternKernel(
        nu=dof,
        lengthscale_prior=cutoff_prior,
        lengthscale_constraint=cutoff_interval
    )
    num_dims = len(grid_size) if grid_bounds is None else None

    gpy_cov = gpytorch.kernels.GridInterpolationKernel(
        gpy_kernel,
        grid_size=grid_size,
        num_dims=num_dims,
        grid_bounds=grid_bounds
    )
    if scale != 1.:
        gpy_cov = gpytorch.kernels.ScaleKernel(
            base_kernel=gpy_cov,
            outputscale_prior=scale_prior,
            outputscale_constraint=scale_interval,
        )
    return gpy_cov


scale, cutoff, dof = 1., 80., 3 / 2
kernel = partial(matern_kernel, scale=scale, cutoff=cutoff, dof=dof)

# %%
shape = (68, 68)
distances = (2., 1.)

pos = np.meshgrid(
    *[d * np.arange(sz) for d, sz in zip(distances, shape)], indexing="ij"
)
pos = torch.asarray(np.stack(pos, axis=-1))
grid_size = shape
mis = np.min(pos.detach().numpy(), axis=tuple(range(len(shape))))
mas = np.max(pos.detach().numpy(), axis=tuple(range(len(shape))))
bumper = 0.1 * (mas - mis)  # Safeguard
grid_bounds = tuple((mi, ma) for mi, ma in zip(mis - bumper, mas + bumper))
pos = pos.reshape(np.prod(shape), -1)

gpy_cov_model = gpytorch_grid_interpolation_cov(
    scale=scale,
    cutoff=cutoff,
    dof=dof,
    grid_size=grid_size,
    grid_bounds=grid_bounds
)

# %%
cov_truth = kernel(distance_matrix(pos.numpy(), pos.numpy()))
cov_gpy = gpy_cov_model(pos).evaluate().detach().numpy()

# %%
imshow_kwargs = {"vmin": cov_truth.min(), "vmax": cov_truth.max()}

fig, axs = plt.subplots(2, 2)
im = axs[0, 0].imshow(cov_truth, **imshow_kwargs)
axs[0, 0].set_title("True Cov")
axs[0, 1].imshow(cov_gpy, **imshow_kwargs)
axs[0, 1].set_title("GPy Cov")
axs[1, 0].axis("off")
im_diff11 = axs[1, 1].imshow(np.abs(cov_truth - cov_gpy))
axs[1, 1].set_title("|True - GPy Cov|")
fig.tight_layout()
fig.colorbar(im, ax=axs[0])
fig.colorbar(im_diff11, ax=axs[1, 1])
plt.savefig("gpytorch_grid_kernel_minimal.png", transparent=True)
```

![gpytorch_grid_kernel_minimal](https://user-images.githubusercontent.com/7541458/165731463-600a67cd-e260-4b1d-a072-a320cdf0b159.png)
